README.md
# App::RegexFileUtils [![Build Status](https://secure.travis-ci.org/plicease/App-RegexFileUtils.png)](http://travis-ci.org/plicease/App-RegexFileUtils)

use regexes with file utils like rm, cp, mv, ln

# SYNOPSIS

Remove all files with a .bak extension:

    % rerm '/\.bak$/'

Change the extension of all files from .jpeg or .JPG (any case) to .jpg

    % remv '/\.jpe?g$/.jpg/i'

Copy all Perl files to a different directory:

    % recp '/\.p[lm]$/' /perl/lib

Create symlinks to .so files so that the symlinks lack a version number

    % reln -s '/\.so\..*$/.so/'

# DESCRIPTION

This distribution provides a version of `rm`, `cp`, `mv` and `ln` with a _re_ 
(as in regular expression) prefix where the file sources can be specified as a regular
expression, or the file source and destination can be specified as a regular expression 
substitution Perl style.  The functionality that this provides can be duplicated with 
shell syntax (typically for loops), but I find these scripts require less typing and 
work regardless of the shell you are using.

The scripts in this distribution do not remove, copy, move or link files directly, 
instead they call the real `rm`, `cp`, `mv` and `ln` programs provided by your
operating system.  You can therefore use any options that they support, for example
the `-i` option will allow you to interactively delete files:

    % rerm -i '/\.bak$/'

# OPTIONS

In addition to any options supported by the underlying operating system, these scripts
will recognize the following options (and NOT pass them to the underlying system utilities).
They are prefixed with `--re` so that they do not interfere with any "real" options.

## --recmd command

Specifies the command to execute.  This is usually determined by Perl's $0 variable.

## --reverbose

Print out the system commands that are actually executed.

## --reall

Include even hidden dot files, like `.profile` and `.login`.

# METHODS

These commands can also be invoked from your Perl script, using this module:

## main

    App::RegexFileUtils->main( $program, @arguments )

For example:

    use App::RegexFileUtils;
    App::RegexFileUtils->main( 'rm', '/\.bak$/' );

# CAVEATS

You will need to enclose many regular expressions in single
quotes '' on the command line as many regular expression characters
have special meanings in shells.

The underlying fileutils command (rm, cp, ln, etc) will be called
for each file operated on, which may be slow if many files match
the regular expression provided.

This was written a long time ago and the code isn't very modern.

Directories with a training slash may be ambiguous with a regex, so
if you want to use a path as a destination instead of a regex, be
sure you do NOT include the trailing slash.  That is:

    # use this:
    % recp /^foo/ /usr/bin
    # NOT this:
    % recp /^foo/ /usr/bin/

# BUNDLED FILES

This distribution comes bundled with `cp`, `ln`, `rm`, `touch`
from the [Perl Power Tools](https://metacpan.org/release/ppt) project.
These are only used if the operating system does not provide these
commands.  This is normally only the case on Windows.  They are individually
licensed separately.

## cp.pl

This program is copyright by Ken Schumack 1999.

This program is free and open software. You may use, modify, distribute
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others from doing the same.

## ln.pl

This program is copyright by Abigail 1999.

This program is free and open software. You may use, copy, modify, distribute,
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others from doing the same.

## rm.pl

Copyright (c) Steve Kemp 1999, skx@tardis.ed.ac.uk

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

## touch.pl

This program is copyright by Abigail 1999.

This program is free and open software. You may use, copy, modify, distribute
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others to do the same.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.08      2017-08-16 15:26:31 -0400
  - Migrate from File::ShareDir to File::ShareDir::Dist

0.07      2016-05-06 05:38:55 -0400
  - Support for MSYS2

0.06      2013-09-16 09:08:27 -0400
  - promote to production

0.05_05   2013-09-14 21:27:53 -0400
  - fix logic in t/retouch.t for when MSWin32 DOES have touch.exe

0.05_04   2013-09-12 12:14:39 -0400
  - Prefer .exe in path over ppt version on MSWin32

0.05_03   2013-09-12 06:29:54 -0400
  - experimental: include my own mv verson for MSWin32 along
    with the ppt programs

0.05_01   2013-09-11 21:49:56 -0400
  - experimental: package ppt versions of ln, cp, rm, touch
    for use on MSWin32 where GNU File/Coreutils (or equivalent)
    are not usually installed.  (mv is missing so I will have
    to write that one myself)

0.05      2013-05-31 07:03:59 America/New_York
  - promote to production

0.04_01   2013-05-29 21:11:34 America/New_York
  - rm dist build dep on [Dist::Zilla::Plugin::MakeMaker::Custom]

0.04      2012-11-26 11:34:01 America/New_York
  - Makefile.PL to fail when cannot find cp mv rm
  - skip ln test if cannot find ln

0.03      2012-11-02 13:54:06 America/New_York
  - Add support for MSWin32 (still need to install GNU FileUtils first)

0.02      2012-10-25 15:25:42 America/New_York
  - Add tests fix bugs.

0.01      2012-10-25 10:55:07 America/New_York
  - initial version


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker ##{ $eumm_version ##};
use File::Spec;

##{ $share_dir_code{preamble} || '' ##}
 
my ##{ 
  my $write_makefile_args = $plugin->write_makefile_args;
  delete $write_makefile_args->{MIN_PERL_VERSION};
   my $makefile_args_dumper = Data::Dumper->new(
     [ $write_makefile_args ],
     [ '*WriteMakefileArgs' ],
   );
   $makefile_args_dumper->Sortkeys( 1 );
   $makefile_args_dumper->Indent( 1 );
   $makefile_args_dumper->Useqq( 1 );
   $makefile_args_dumper->Dump;
##}

my $sep = $^O eq 'MSWin32' ? ';' : ':';
my $ext = $^O =~ /^(MSWin32|cygwin)$/ ? '.exe' : '';
my %found;
foreach my $path (split $sep, $ENV{PATH})
{
  foreach my $program (qw( mv cp rm ))
  {
    my $exe = File::Spec->catfile($path, "$program$ext");
    if(-x $exe)
    {
      $found{$program} = 1;
    }
  }
}

foreach my $program (qw( mv cp rm ))
{
  warn "not found: $program" unless $found{$program};
}

unless($found{mv} && $found{cp} && $found{rm})
{
  warn "this distribution requires GNU Coreutils (mv, cp, rm and ln), or equivalent";
  if($^O eq 'MSWin32')
  {
    warn "can be downloaded from the GnuWin32 project: http://gnuwin32.sourceforge.net/";
  }
  exit 2;
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}
 
delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };
 
WriteMakefile(%WriteMakefileArgs);
 
##{ $share_dir_code{postamble} || '' ##}



inc/MakeMaker.pm
package inc::MakeMaker;

use Moose;
use namespace::autoclean;

with 'Dist::Zilla::Role::InstallTool';

my $checks;

sub setup_installer
{
  my($self) = @_;
  
  my($makefile) = grep { $_->name eq 'Makefile.PL' } @{ $self->zilla->files };
  
  my $content = $makefile->content;
  
  unless($checks)
  {
    $checks = do { local $/; <DATA> };
  }
  
  if($content =~ s{(WriteMakefile\()}{$checks$1}m)
  {
    $makefile->content($content);
    $self->zilla->log("Modified Makefile.PL with extra checks");
  }
  else
  {
    $self->zilla->log_fatal("unable to update Makefile.PL");
  }
}

1;

__DATA__

my $sep = $^O eq 'MSWin32' ? ';' : ':';
my $ext = $^O =~ /^(MSWin32|cygwin)$/ ? '.exe' : '';
my %found;
foreach my $path (split $sep, $ENV{PATH})
{
  foreach my $program (qw( mv cp rm ))
  {
    my $exe = File::Spec->catfile($path, "$program$ext");
    if(-x $exe)
    {
      $found{$program} = 1;
    }
  }
}

foreach my $program (qw( mv cp rm ))
{
  warn "not found: $program" unless $found{$program};
}

# On Windows we use ppt implementation of some of these
# if they are not found i nthe path.
$found{cp} = $found{rm} = $found{mv} = $found{touch} = 1 if $^O eq 'MSWin32';

unless($found{mv} && $found{cp} && $found{rm})
{
  warn "this distribution requires GNU Coreutils (mv, cp, rm and ln), or equivalent";
  if($^O eq 'MSWin32')
  {
    warn "can be downloaded from the GnuWin32 project: http://gnuwin32.sourceforge.net/";
  }
  exit 2;
}





lib/App/RegexFileUtils/MSWin32.pm
package App::RegexFileUtils::MSWin32;

# ABSTRACT: MSWin32 specific code for App::RegexFileUtils
# VERSION

package
  App::RegexFileUtils;

use strict;
use warnings;
use File::ShareDir qw( dist_dir );
use File::Basename qw( dirname );
use File::Spec;

warn "only needed on MSWin32" unless $^O eq 'MSWin32';

my $path;

sub share_dir
{
  unless(defined $path)
  {
    if(defined $App::RegexFileUtils::MSWin32::VERSION && __FILE__ =~ /blib/)
    {
      $path = File::Spec->catdir(
        dirname(__FILE__), 
        File::Spec->updir,
        File::Spec->updir,
        File::Spec->updir,
        File::Spec->updir,
        'share',
      );
      undef $path unless $path && -d "$path/ppt";
    }
    
    eval {
      if(defined $App::RegexFileUtils::MSWin32::VERSION)
      {
        $path = dist_dir('App-RegexFileUtils');
      }
      undef $path unless $path && -d "$path/ppt";
    };
    
    unless(defined $path)
    {
      $path = File::Spec->catdir(
        dirname(__FILE__),
        File::Spec->updir,
        File::Spec->updir,
        File::Spec->updir,
        'share',
      );
    }
    
    die 'can not find share directory' unless $path && -d "$path/ppt";    
  }
  
  $path;
}

sub fix_path
{
  my($class, $cmd) = @_;

  foreach my $path (split /;/, $ENV{PATH})
  {
    return if -x File::Spec->catfile($path, $cmd->[0] . '.exe');
  }

  $cmd->[0] = File::Spec->catfile(
    App::RegexFileUtils->share_dir,
    'ppt', $cmd->[0] . '.pl',
  );
  unshift @$cmd, $^X;
}

1;


lib/App/RegexFileUtils.pm
package App::RegexFileUtils;

use strict;
use warnings;
use File::Spec;
use File::ShareDir::Dist qw( dist_share );
use File::Which qw( which );

# ABSTRACT: use regexes with file utils like rm, cp, mv, ln
# VERSION

=head1 SYNOPSIS

Remove all files with a .bak extension:

 % rerm '/\.bak$/'

Change the extension of all files from .jpeg or .JPG (any case) to .jpg

 % remv '/\.jpe?g$/.jpg/i'

Copy all Perl files to a different directory:

 % recp '/\.p[lm]$/' /perl/lib

Create symlinks to .so files so that the symlinks lack a version number

 % reln -s '/\.so\..*$/.so/'

=head1 DESCRIPTION

This distribution provides a version of C<rm>, C<cp>, C<mv> and C<ln> with a I<re> 
(as in regular expression) prefix where the file sources can be specified as a regular
expression, or the file source and destination can be specified as a regular expression 
substitution Perl style.  The functionality that this provides can be duplicated with 
shell syntax (typically for loops), but I find these scripts require less typing and 
work regardless of the shell you are using.

The scripts in this distribution do not remove, copy, move or link files directly, 
instead they call the real C<rm>, C<cp>, C<mv> and C<ln> programs provided by your
operating system.  You can therefore use any options that they support, for example
the C<-i> option will allow you to interactively delete files:

 % rerm -i '/\.bak$/'

=head1 OPTIONS

In addition to any options supported by the underlying operating system, these scripts
will recognize the following options (and NOT pass them to the underlying system utilities).
They are prefixed with C<--re> so that they do not interfere with any "real" options.

=head2 --recmd command

Specifies the command to execute.  This is usually determined by Perl's $0 variable.

=head2 --reverbose

Print out the system commands that are actually executed.

=head2 --reall

Include even hidden dot files, like C<.profile> and C<.login>.

=head1 METHODS

These commands can also be invoked from your Perl script, using this module:

=head2 main

 App::RegexFileUtils->main( $program, @arguments )

For example:

 use App::RegexFileUtils;
 App::RegexFileUtils->main( 'rm', '/\.bak$/' );

=head1 CAVEATS

You will need to enclose many regular expressions in single
quotes '' on the command line as many regular expression characters
have special meanings in shells.

The underlying fileutils command (rm, cp, ln, etc) will be called
for each file operated on, which may be slow if many files match
the regular expression provided.

This was written a long time ago and the code isn't very modern.

Directories with a training slash may be ambiguous with a regex, so
if you want to use a path as a destination instead of a regex, be
sure you do NOT include the trailing slash.  That is:

 # use this:
 % recp /^foo/ /usr/bin
 # NOT this:
 % recp /^foo/ /usr/bin/

=head1 BUNDLED FILES

This distribution comes bundled with C<cp>, C<ln>, C<rm>, C<touch>
from the L<Perl Power Tools|https://metacpan.org/release/ppt> project.
These are only used if the operating system does not provide these
commands.  This is normally only the case on Windows.  They are individually
licensed separately.

=head2 cp.pl

This program is copyright by Ken Schumack 1999.

This program is free and open software. You may use, modify, distribute
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others from doing the same.

=head2 ln.pl

This program is copyright by Abigail 1999.

This program is free and open software. You may use, copy, modify, distribute,
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others from doing the same.

=head2 rm.pl

Copyright (c) Steve Kemp 1999, skx@tardis.ed.ac.uk

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head2 touch.pl

This program is copyright by Abigail 1999.

This program is free and open software. You may use, copy, modify, distribute
and sell this program (and any modified variants) in any way you wish,
provided you do not restrict others to do the same.

=cut

sub main {
  my $class = shift;
  my $mode = shift;
  my $appname = $mode;
  $mode =~ s!^.*/!!;
  my @args = @_;
  
  my @options = ();
  my $verbose = 0;
  my $do_hidden = 0;
  my $re;
  my $sub;
  my $modifiers;
  my $modifiers_match = '';
  my $tr;
  my $static_dest = 0;
  
  while(defined($args[0]) && $args[0] =~ /^-/) {
    my $arg = shift @args;
  
    if($arg =~ /^--recmd$/) {
      $mode = shift @args;
      $mode =~ s!^.*/!!;
    } elsif($arg =~ /^--reverbose$/) {
      $verbose = 1;
    } elsif($arg =~ /^--reall$/) {
      $do_hidden = 1;
    } else {
      push @options, $arg;
    }
  }
  
  my $dest = pop @args;
  
  unless(defined $dest) {
    print STDERR "usage: $appname [options] [source files] /pattern/[substitution/]\n";
    print STDERR "       $appname [options] /pattern/ /path/to/destination\n";
    print STDERR "\n";
    print STDERR "--recmd [command]      change the behavior of the tool\n";
    print STDERR "--verbose              print commands before they are executed\n";
    print STDERR "--reall                include hidden (so called `dot') files\n";
    print STDERR "\n";
    print STDERR "all other arguments are passed to the system tool\n";
    exit;
  }
  
  my $orig_mode = $mode;
  $mode =~ s/^re//;
  
  my %modes = (
    'mv'    => 'mv',
    'move'    => 'mv',
    'rename'  => 'mv',
    'cp'    => 'cp',
    'copy'    => 'cp',
    'ln'    => 'ln',
    'link'    => 'ln',
    'symlink'  => 'ln',
    'rm'    => 'rm',
    'remove'  => 'rm',
    'unlink'  => 'rm',
    'touch'    => 'touch',
  );
  
  unshift @options, '-s' if $mode eq 'symlink';
  
  $mode = $modes{$mode};
  unless(defined $mode) {
    print STDERR "unknown mode $orig_mode\n";
    exit;
  }
  
  my $no_dest = 0;
  if($mode eq 'touch' || $mode eq 'rm') {
    $no_dest = 1;
  }
  
  if($dest =~ m!^(s|)/(.*)/(.*)/([ig]*)$!) {
    $re = $2;
    $sub = $3;
    $modifiers = $4;
    $modifiers_match = 'i' if $modifiers =~ /i/;
  
    if($no_dest) {
      print STDERR "substitution `$mode' doesn't make sense\n";
      exit;
    }
  
  }
  
  elsif($dest =~ m!tr/(.*)/(.*)/$!) {
    $tr = $1;
    $sub = $2;
  
    if($no_dest) {
      print STDERR "translation `$mode' doesn't make sense\n";
    }
  }
  
  elsif($dest =~ m!^(m|)/(.*)/([i]*)$!) {
    $re = $2;
    $modifiers = $3;
    $modifiers_match = $3;
  }
  
  elsif(-d $dest) {
    my $src = pop @args;
    if($src =~ m!^(m|)/(.*)/([i]*)$!) {
      $static_dest = 1;
      $re = $2;
      $modifiers = $3;
      $modifiers_match = $3;
    } else {
      die "source is not a regex";
    }
  }
  
  else {
    die "destination is not a directory or a regex";
  }
  
  my @files = @args;
  
  if(@files ==0) {
    opendir(DIR, '.') || die "unable to opendir `.' $!";
    @files = readdir(DIR);
    closedir DIR;
  }
  
  for(@files) {
    next if /^\./ && !$do_hidden;
    next unless eval "/$re/$modifiers_match" || defined $tr;
    my $old = $_;
    my $new = $old;
    
    my @cmd = ($mode, @options, $old);
    
    if(defined $tr) {
      eval "\$new =~ tr/$tr/$sub/";
    } elsif(defined $sub) {
      eval "\$new =~ s/$re/$sub/$modifiers";
    } elsif($static_dest) {
      $new = $dest;
    } else {
      if($no_dest) {
        $new = '';
      } else {
        $new = '.';
      }
    }
    
    push @cmd, $new unless $no_dest;
    print "% @cmd\n" if $verbose;
    
    __PACKAGE__->_fix_path(\@cmd);
    
    system @cmd;

    if ($? == -1) {
      print STDERR "failed to execute: $!\n";
      exit 2;
    } elsif ($? & 127) {
      print STDERR "child died with signal ", $? & 127, "\n";
    } elsif($? >> 8) {
      print "child exited with value ", $? >> 8, "\n";
    }
  }
}

use constant _share_dir => do {
  my $path;
  $path = dist_share('App-RegexFileUtils');
  die 'can not find share directory' unless $path && -d "$path/ppt";    
  $path;
};

sub _fix_path
{
  my($class, $cmd) = @_;

  return unless $^O eq 'MSWin32';

  return if which($cmd->[0]);

  $cmd->[0] = File::Spec->catfile(
    App::RegexFileUtils->_share_dir,
    'ppt', $cmd->[0] . '.pl',
  );
  unshift @$cmd, $^X;
}

1;
